[RFC/WIP] NVPTX/Cuda backend support in Julia

vchuravy opened the pull request:
## Goal

The long term goal is to enable support for compiling Julia to different accelerator backends like NVPTX and SPIR and in the future SPIR-V and HSAIL. This PR is a rebase of the work done by @maleadt at [maleadt/master](https://github.com/maleadt/Julia) onto the current Julia master. 
It enables the CUDA backend and in combination with [CUDA.jl](https://github.com/vchuravy/CUDA.jl/tree/vc/fixes) it can compile simple CUDA kernels written in Julia as described in the [blog of maleadt](http://blog.maleadt.net/2015/01/15/julia-cuda)

This is still in the early stage and I would like to invite comments and recommendations for how to proceed further. 
## Components
### Target

edf0693 enables the CUDA target, target selection with `@target` and code compilation. This is the center point of this PR. I would like to discuss the approach taken and how we could scale that to several different backends.
### llvmcall

10434e6 enhances the capabilities of llvmcall as described in #8740. This is necessary to be able to call the NVPTX [intrinsics](https://groups.google.com/forum/#!topic/julia-dev/doQUqQabC8A). 
### bitcasts

Julia ignores address space and to make backends work that use different address space our bitcast needs to respect the address space. See  #9423 for more information on the matter.
## TODO:
- [ ] Tests
  - Should the test compare IR output or can we set up a test architecture with [gpuocelot](https://github.com/maleadt/gpuocelot)
- [ ] https://github.com/maleadt/julia/blob/master/TODO.md
  - Refuse GC allocations
  - Refuse calling unknown functions
  - Allow codegen of ordinary arrays (`auto_unbox`?)
  - Allow multiple active codegen contexts. We could possibly generalize this,
    i.e. have 1 single context for host codegen, multiple ones (stack?) for PTX
    code, and select one in `emit_function` based on the active/parent target.
  - use `@target` to refuse cross-target calls
  - Implicit `return nothing`, and check for non-void kernels
  - Remove the old codegen exceptions for Array values (in the past we lowered
    normal Arrays to raw pointers, now we use an immutable `CuDeviceArray` for
    that purpose).
## Credits

Most of the work has been done by @maleadt and `Pieter Verstraete` see https://github.com/maleadt/julia/compare/master for a proper history. My part in the work was mostly squashing and rebasing the code onto upstream Julia, since @maleadt announced that he won't have time to work on this (see [post on julia-dev](https://groups.google.com/forum/#!topic/julia-dev/doQUqQabC8A)). I plan to continue working on this to get SPIR/SPIR-V support. Thanks to @SimonDanisch for helping with testing and motivating me. 
## How to try this

If you don't have a CUDA capable card you will need to use `gpuocelot` like described in http://blog.maleadt.net/2015/01/15/julia-cuda. Substitute CUDA.jl with my fork at https://github.com/vchuravy/CUDA.jl/tree/vc/fixes

Commit:
Enable nvptx compilation in llvm

Commit:
Make most bitcasts preserve the address space property.

Commit:
Extend llvmcall with support for entering declarations.

maleadt committed:
introduces the cuda target and the `@target` macro

vchuravy committed:
adjust llvmcall test to tuple changes

vchuravy committed:
support llvm35 and prevent seqfault if NVVMIR_LIBRARY_DIR is not present

vchuravy committed:
add second default location

vchuravy commented:
The travis errors stem from me using LLVM 3.5.0. I will test with LLVM 3.3.0

vchuravy committed:
Error only needs to be defined on LLVM35

rollingthunder commented:
Over the past several months, I have been working on integrating the HSA Runtime with julia for my master's thesis. The goal is to provide a wrapper library similar to OpenCL.jl and CUDA.jl to simplify usage of HSA from julia, as well as to integrate compilation of julia kernel functions to HSAIL/BRIG Targets. 

So far, I have mostly concentrated on implementing the [HSA.jl](https://github.com/rollingthunder/HSA.jl) Runtime Wrapper and integrating the julia build system with the [HSA development version of LLVM](https://github.com/HSAFoundation/HLC-HSAIL-Development-LLVM)

I am now starting to implement the compilation of julia functions to HSAIL.

It looks like the compilation to HSAIL is fairly straightforward, if what you have as an input is SPIR 2.0.
SPIR could also be reused in targeting OpenCL. For these reasons, the plan for my prototype is to implement the (most important parts of the) SPIR Target and, from there, the compilation to HSAIL.

Since you are already looking to build a scalable infrastructure to compile to multiple device targets, I wanted to let you know of my work now. 

Maybe we could come up with something that I can already use in my implementation.

vchuravy commented:
@rollingthunder That is great to hear. So this PR should give you a hint how to integrate an additional target into the Julia backend. Also you might find the modified 'llvmcall` useful.

Why do you need to use the SPIR target? I would think that you enable HSAIL as a target similar to CUDA in this PR and don't need the step in between.

tkelman commented:
Is there a feature-detection macro from LLVM to determine whether this target is enabled? I imagine some system copies of llvm might not enable it?

vchuravy commented:
@tkelman There is `llvm/Config/Targets.def`, but I haven't found out how to turn that into a simple `#ifdef`. So you can do stuff like below:

````
enum llvm_targets {
#define LLVM_TARGET(TargetName) TargetName,
#include "llvmc/Config/Targets.def"
#undef LLVM_TARGET
};
````

rollingthunder commented:
The reason I plan to go through SPIR on the way to HSAIL is that the SPIR intrinsics are well documented and there is a mapping from SPIR intrinsics to HSAIL intrinsics (builtins-hsail.ll). 

HSAIL LLVM intrinsics are not really documented anywhere as far as I can tell.

There may well be HSAIL intrincics that don't have a mapping from SPIR, but I would rather add this subset using llvmcall later than use only HSAIL intrinsics from the beginning. 

Also, this way you could conceivably write one kernel function in julia (@target spir maybe) and use that both on OpenCL and on HSA.

tkelman commented:
This'll need to do something to allow Julia to still build correctly with these features disabled if it's linking against a copy of LLVM that doesn't support this target. I don't see how this could realistically be merged to master otherwise.

vchuravy commented:
I agree, I will see what I can come up with.

On Fri, 5 Jun 2015 at 20:25 Tony Kelman <notifications@github.com> wrote:

> This'll need to do something to allow Julia to still build correctly with
> these features disabled if it's linking against a copy of LLVM that doesn't
> support this target. I don't see how this could realistically be merged to
> master otherwise.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/11516#issuecomment-109262604>.
>

vchuravy committed:
initialize all available llvm targets.

Clients that require a particular target like `NVPTX` have to ensure
that the target is available in the linked llvm and set up the
appropriate context. See `jl_init_codegem_ptx` as an example.

vchuravy commented on src/codegen.cpp:
```diff
@@ -681,6 +712,124 @@ static Function *to_function(jl_lambda_info_t *li)
+// TODO: a lot of duplication with to_function and init_julia_llvm_env
+// TODO: invoke by call()ing a @target("ptx") function
+extern "C" DLLEXPORT
+const jl_value_t * jl_to_ptx() {
```
So after taking a look at https://github.com/Keno/DIDebug.jl it should be possible to write `jl_to_ptx` in terms of `CXX.jl` and keep it out of Julia base.

This would slim down this PR to `@target` (and maybe `@kernel`), some special cases for NVPTX, and more or less independent PRs for `preserving_bitcast` and `llvmcall`

malmaud commented:
@vchuravy I'm excited to see this line of work pushed forward - are you still interested in working on this PR?

vchuravy commented:
@malmaud Yes I am still interested. I am just waiting for v0.4 to be released and the codegen changes merged.

ViralBShah commented:
A lot of work is pending on the new codegen changes of @vtjnash, and that is probably one of the first things that will happen after 0.4 (or perhaps even on branching 0.4).

ViralBShah commented:
A few folks have mentioned this PR to me recently. What keeps us from getting it merged now?

vchuravy commented:
@ViralBShah First of all #11604 would need to be rebased and merged (if the implementation is unontroversial) and after that #9423 would need to be rebased and extended (lots of changes happened since then).

After these changes landed this can move forward, the general questions that would need to be answered whether this should live inside Julia base or if backends should be implemented in associated packages. In the future we would like to add more and more backends and target specific logic should imho not be part of Julia. The only exception I would make is for a very general backend like SPIR-V that could potentially transcend target specific backends.

I hope that explains the current situation. I could rebase #11604 over the weekend.

PS: This also depends on switching to llvm 3.5+

ViralBShah commented:
Thank you for outlining what needs to be done. I do like the idea of adding additional backends as packages. I guess we are not too far from llvm 3.7 now, but @keno and others know best.

vchuravy commented:
We would need some infrastructure in base though. I remeber @jakeb
<https://github.com/jakebolewski>olewsk <https://github.com/jakebolewski>i
<https://github.com/jakebolewski>  talking about the need for a Julia llvm
interface in  https://github.com/JuliaGPU/OpenCL.jl/issues/29, but maybe
that is more feasible now with CXX.jl

On Wed, 28 Oct 2015 at 23:22 Viral B. Shah <notifications@github.com> wrote:

> Thank you for outlining what needs to be done. I do like the idea of
> adding additional backends as packages. I guess we are not too far from
> llvm 3.7 now, but @keno <https://github.com/keno> and others know best.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/11516#issuecomment-151860246>.
>

maleadt commented:
I've very recently also been spending some time on the back-end again (see [maleadt/julia](https://github.com/maleadt/julia/commits/master) and [maleadt/CUDA.jl](https://github.com/maleadt/CUDA.jl/commits/master)), but some code generation changes are holding me back. I'm currently trying to get the commit before the mega codegen restructure working, on LLVM 3.7.

vchuravy commented:
Tim, nice to hear from you. What do you think would be necessary for having
such backends outside of Julia proper?

On Wed, 28 Oct 2015 at 23:37 Tim Besard <notifications@github.com> wrote:

> I've very recently also been spending some time on the back-end again (see
> maleadt/julia <https://github.com/maleadt/julia/commits/master> and
> maleadt/CUDA.jl <https://github.com/maleadt/CUDA.jl/commits/master>), but
> some code generation changes are holding me back. I'm currently trying to
> get the commit before the mega codegen restructure working, on LLVM 3.7.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/11516#issuecomment-151866064>.
>

maleadt commented:
Despite not being happy with the current `if (ctx.target == PTX)` proliferation, I haven't given it too much thought. We obviously would need a decent Julia interface to LLVM (would the C interface suffice here?). Maybe `emit_function` could conditionally (say, based on `@target`) try and load registered `emit_function_for_$TARGET` functions? I guess this would depend on the codegen utility functions being more modular than they are right now, and `emit_function` to be somewhat less monolithic, both of which are quite an undertaking in itself.

maleadt commented:
Specific to the CUDA back-end, it would also be useful to have Julia support for pointer address spaces. How feasible would it be to parametrize `Ptr` in terms of the address space, with the default parameter value being `0` to preserve current behaviour? I had a go at it ages ago, but messing with `Ptr` seemed to break... a lot.

vchuravy commented:
@rollingthunder You have been working on an HSA/SPIR backend. Would you mind sharing your experiences and what you would need from base?

vchuravy commented:
SPIR-V 1.0 was just released [1] and Khronos provides some LLVM extensions that translate OCL2.0 and OCL2.1 to SPIR-V [2]. When I find time I will focus on address-space aware bitcast and a Julia to SPIR-V translator. 


[1] https://www.khronos.org/spir
[2] https://github.com/KhronosGroup/SPIRV-LLVM

rollingthunder commented:
@vchuravy
Some of the major issues with HSAIL were:
- The lack of support for address spaces in julia.  
For example, when we were emitting an HSAIL kernel, we would have liked to force its signature to use pointer arguments in the global address space and then have the codegen use these types when generating the function body. 
That didn't work, because there are places in the cg where (even with the preserving_bitcast PR) non-generic AS values do bad things.
- A way to prevent the boxing of values and the associated GC frame emission.  
Since we cannot translate that to device code, any kernel that triggers boxing cannot be compiled.
Just suppressing any GC frame code might already help, though obviously the caller of the GPU kernel would need to take care of keeping a reference to memory that is being used. 

Other things like how to handle the device code generators need solving too, but it seems that is already being talked about: JuliaGPU/meta#2

maleadt commented:
From what I've seen about gpucc, they've developed an address space inference pass to propagate AS information from e.g. the function's entry point, so that might be solvable without much changes to Julia itself. Still, decent language support might be interesting.

vchuravy commented:
Maybe instead of changing the pointer interface we could adapt the Ref interface?
